### PR TITLE
Feature/red experience page

### PR DESCRIPTION
**Pull request by JohnJAMoore**
Add red experience page

Issue: Scottish-Natural-Heritage/Licensing#342

### CODE REVIEW

File: src/controllers/red-experience.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * database and its validation rules.
+ *
+ * @param {object} body The incoming request's body.
+ * @returns {any} A json object that's just got our cleaned up fields on it.
+ */
+const cleanInput = (body) => {
+  return {
+    // Ensure the user has entered integers
+    cleanRedExperienceYears: body.redExperienceYears === undefined ? undefined : body.redExperienceYears.trim(),
+    cleanRedControlNumber: body.redControlNumber === undefined ? undefined : body.redControlNumber.trim()
```
Comment (MikeCoats) [inline src/controllers/red-experience.js]: Either...
1. That comment is wrong.
2. You're not using the [util function we wrote to clean integers](https://github.com/Scottish-Natural-Heritage/fit-and-competent/blob/develop/src/controllers/_util.js#L8).

File: src/controllers/red-experience.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    request.session.redExperienceYears = cleanRedExperienceYears;
+    request.session.redControlNumber = cleanRedControlNumber;
+
+    if (request.session.redExperienceYears) {
+      request.session.redExperienceYearsInvalidError = new RegExp(/\D/).test(cleanRedExperienceYears);
+    }
+
+    if (request.session.redControlNumber) {
+      request.session.redControlNumberInvalidError = new RegExp(/\D/).test(cleanRedControlNumber);
+    }
```
Comment (MikeCoats) [inline src/controllers/red-experience.js]: If you do end up using the util/cleanInt function, these tests won't work correctly any more and should be done on the trimmed, user supplied strings.

**Review by MikeCoats — CHANGES_REQUESTED** [review]
One little thing to tidy up, but it's nearly 100%.

A quick spell check says there's nothing obvious on that front! 😁
image

File: src/controllers/red-experience.js [hunk 3]
```diff
@@ -28,21 +28,21 @@ const redExperienceController = (request) => {
...
-    if (request.session.redControlNumber) {
-      request.session.redControlNumberInvalidError = new RegExp(/\D/).test(cleanRedControlNumber);
+    if (request.body.redControlNumber) {
+      request.session.redControlNumberInvalidError = new RegExp(/\D/).test(request.body.redControlNumber.trim());
     }
 
-    if (request.session.redExperienceYears === '' && request.session.redControlNumber === '') {
+    if (request.body.redExperienceYears.trim() === '' && request.body.redControlNumber.trim() === '') {
       request.session.redEmptyError = true;
     }
```
Comment (MikeCoats) [inline src/controllers/red-experience.js]: I think this test should be expanded.

Currently it checks if BOTH fields are empty, but then only tells the user about one error.

image

If the user then corrects that one error, they're taken on the forward path, not back here to correct the second missing field.

There's also a possibility that `request.body.redExperienceYears` or `request.body.redControlNumber` could be undefined at this point, so you'll get `Cannot read property 'trim' of undefined` errors.

I suggest that this test is split in two and each part is embedded inside the two tests above it.

File: src/views/red-experience.njk [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        errorMessage: {
+          html: redExperienceYearsErrors
+        } if model.redExperienceYearsInvalidError or model.redEmptyError
+      }) }}
+
+      <h4 class="govuk-heading-s">
+        What is the average number of red deer you have controlled per year over the last 5 years?
+      </h4>
+      {{ govukInput({
+        classes: "govuk-input--width-4",
```
Comment (MikeCoats) [inline src/views/red-experience.njk]: This is getting really petty now, but I don't like the width of these controls. 😟

The two digit one looks really out of place to me. And the fact they're different widths seems awkward too.

According to the GDS example, they've used a 10 width input for an 8 digit number, https://design-system.service.gov.uk/components/text-input/#number-input-example, so could we try a 6-width input for both of them and see what it looks like?

File: src/views/red-experience.njk [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      <ul class="govuk-list">
+        <li>01463 725 364</li>
+        <li>Monday to Friday, 10am to 4pm (except public holidays)</li>
+      </ul>
+      <p class="govuk-body">
+        <a class="govuk-link" rel="noreferrer noopener" target="_blank" href="https://www.gov.uk/call-charges">Find out about call charges&nbsp;(opens in new tab)</a>
+      </p>
+      <a class="govuk-link govuk-link--no-visited-state govuk-!-font-size-16" href="#top-of-page">
+        Back to top
+      </a>
```
Comment (MikeCoats) [inline src/views/red-experience.njk:136]: I also just noticed this.

I think you've just copied and pasted from a previous page and included too much. I've done a quick check and it's on quite a few pages, so I'll make a new card to address all the pages in one go.

**Review by MikeCoats — CHANGES_REQUESTED** [review]
I think there's some work to do in one of the errors, and maybe a little 'designer-y' work on the size of the inputs.

File: src/views/red-experience.njk (hunk 4 above)
Comment (JohnJAMoore) [reply src/views/red-experience.njk]: no worries ill change this to a width of 6

File: src/controllers/red-experience.js (hunk 3 above)
Comment (JohnJAMoore) [reply src/controllers/red-experience.js]: "If the user then corrects that one error, they're taken on the forward path, not back here to correct the second missing field."

So I asked jimmy about the validation on these experience pages as there was a potential scenario where a person may be returning to stalking deer but hasn't controlled any in the last 5 years so wouldn't have to fill out the second box the response I received from him about this is below:

"I would say they can proceed but this should be flagged, no automatic approval"

Comment (MikeCoats) [reply src/controllers/red-experience.js]: > hasn't controlled any in the last 5 years

There's a difference between 0 and undefined. 0 means 'none', undefined means '🤷🏻‍♀️'. The correct answer for the user here is...

> **What is the average number of red deer you have controlled per year over the last 5 years?**
> `0`


'🤷🏻‍♀️' means you cannot proceed. [We don't ask questions we don't want answers for](https://design-system.service.gov.uk/patterns/question-pages/#when-to-use-this-pattern).

>You should make sure you know why you’re asking every question and only ask users for information you really need.

I's strongly argue against it, but if you really want to make this field optional, GDS has this to say.

> If you ask for optional information, mark the labels of optional fields with ‘(optional)’.

**Review by MikeCoats — APPROVED** [review]
Yes, this is all good now. Nice one!

**Pull request merged**